### PR TITLE
Add callbacks, history logging and performance improvements to sgd.

### PR DESCRIPTION
**Pull request by samjmolyneux**
- By using `jax.jit`, and calculating the gradient and objective in a single pass with `jax.value_and_grad`, we get big speed boosts to sgd. The new time taken is approximately $\frac{1}{7}\text{th}$ of the original for the two normal example on my machine. <p></p>
- New `history_logging_interval` parameter for the `stochastic_gradient_descent` function allows the user to enable or disable logging of the optimisation history. The interval determines how frequently the history is logged. This makes it easier to debug optimisations and make decisions about hyperparameters. <p></p>
- To make history logging work, we add a dataclass `IterationResult` which is analogous to `SolverResult`. However, `IterationResult` uses `frozen=True` to allow for dataclass updates each iteration. Using a dataclass ensures backward compatibility for callbacks if a new attribute is logged.  <p></p>
- New `callbacks` parameter for the `stochastic_gradient_descent` function allows the user to set a list of callback functions as is standard in optimisation loops. In future, the callbacks can be used for early stopping or live plotting of the results. As an example, we include a useful `tqdm` callback that displays a progress bar for the iterations and displays the current objective value.   <p></p>
- We also add the following tests:
- - `test_normalise_callbacks`: Tests that `_normalise_callbacks` does validation and casts valid types to `list[Callable[IterationResult], None]` .
- - `test_sgd_history_logging_intervals`: Tests that the correct iterations are logged for different intervals and that the correct associated obj, fn_args and grad are too for sgd.
- - `test_callback_invocation`: Tests that sgd callbacks are called in the correct order with the correct `IterationResult`.
- - `test_invalid_callback`: Tests that sgd will raise an error if given an invalid callback. 
- - `test_logging_or_callbacks_affect_sgd_convergence`: Tests that various combinations of callbacks and logging intervals all result in the same convergence behaviour and thus all have the same final obj, fn_args, grad etc.

### CODE REVIEW

**Review by willGraham01 — COMMENTED** [review]
All-in-all I like these changes, and these are useful features that we could do with adding. Callbacks in particular should be very useful from a development/debugging perspective too,

Most of my comments relate to the design decisions for the code, considering what's in the rest of the codebase. Namely I think we can do some code recycling in places, and we tend to write our tests in a particular format (though the test cases provided are good).

Also, there are only two commits on this branch (one for codebase changes, one for tests). In general, don't be afraid to use more granular commits in your PRs (just take a look at how long the other PRs are!) - we use squash merges anyway, so everything gets condensed into a single commit on `main` anyway. And it's good to be able to roll things back.

File: src/causalprog/solvers/iteration_result.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    iters: int
+    obj_val: npt.ArrayLike
+
+    iter_history: list[int] = field(default_factory=list)
+    fn_args_history: list[PyTree] = field(default_factory=list)
+    grad_val_history: list[PyTree] = field(default_factory=list)
+    obj_val_history: list[npt.ArrayLike] = field(default_factory=list)
+
+
+def _update_iteration_result(
```
[the remark this answers is not in the document]
Comment (samjmolyneux) [reply src/causalprog/solvers/iteration_result.py:39]: > Any reason why this isn't a method of the `IterationResult` class? I expected this to be something like `IterationResult.update` (and it's currently written in that form too -> swap `iter_result` to `self`).

Nope. It should be. 

> Related, any reason for why `history_logging_interval` is an argument that we pass in, rather than an attribute that's set at creation time (I guess we could want dynamic logging which we wouldn't get with a fixed attribute, but is that a common enough use-case to design around?). It also means that we could just check `history_logging_interval > 0` once at creation time, and not do it every time in the method.

No. This is a good point.

File: src/causalprog/solvers/sgd.py [hunk 2]
```diff
@@ -82,28 +96,49 @@ def stochastic_gradient_descent(
...
-        objective_value = objective(current_params)
-        gradient_value = gradient(current_params)
+        _update_iteration_result(
+            iter_result,
+            current_params,
+            gradient_value,
+            _,
+            objective_value,
+            history_logging_interval,
+        )
```
[the remark this answers is not in the document]
Comment (samjmolyneux) [reply src/causalprog/solvers/sgd.py:130]: Yeah, because of the `iters_used = _` line, I wasn't sure if this was a convention you were using, so I didn't want to change it without checking.

File: src/causalprog/solvers/solver_callbacks.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+def _normalise_callbacks(
+    callbacks: Callable[[IterationResult], None]
+    | list[Callable[[IterationResult], None]]
+    | None = None,
+) -> list[Callable[[IterationResult], None]]:
+    if callbacks is None:
+        return []
+    if callable(callbacks):
+        return [callbacks]
+    if isinstance(callbacks, list) and all(callable(cb) for cb in callbacks):
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply src/causalprog/solvers/solver_callbacks.py]: > Or maybe it is best to just do iterables without type checks and beg forgiveness like you say

I think go "try and beg forgiveness". As we've established, someone would _really_ have to actively go out of their way to create & pass something in here that could mess stuff up.

**Review by willGraham01 — APPROVED** [review]
Looks like everything has been addressed and was waiting on my approval.

Unless you have any pending local changes @samjmolyneux, hit merge when you're ready.
